Add public docs for device trust

codingllama opened the pull request:
Add the following public documentation:

* Device Trust guide under access control
* Device Trust configuration reference
* `tctl devices` command reference
* `tsh device enroll` command reference

https://github.com/gravitational/teleport.e/issues/514

codingllama committed:
Add device trust documentation

codingllama committed:
Document `device_trust` authentication section

codingllama committed:
Document the `tsh device enroll` command

codingllama committed:
Document `tctl devices` commands

codingllama committed:
Link to Device Trust from Access Control pages

zmb3 commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+- SSH access enforcement
+- Database access enforcement
+- Kubernetes access enforcement
+- macOS trusted devices
```
The last one is more of a system requirement than a Teleport feature.

I would say "the device trust preview requires macOS and supports the following Teleport features:"

codingllama replied:
Rephrased as "The device trust preview works only with macOS devices and supports the following Teleport features". Let me know if you prefer your wording.

zmb3 commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+Support for other operating systems and access features is planned for upcoming
+Teleport versions.
+
+## Concepts
```
Some of the feedback we've seen from the community lately is a desire to see more "how it works" docs instead of framing everything as a guide.

I wonder if it makes sense to pull some of these concepts out into a separate intro page for people who want to understand the feature but don't care (or aren't ready) to read through a step-by-step guide on how to set it up.

codingllama replied:
I thought briefly about a separate page too. Maybe we could move this part under architecture?

The reason I added it here is that, unlike other features, device trust isn't self-explanatory. Even a weakly-opinionated system like our has to make some calls on how devices are registered, enrolled, etc, so it felt necessary to explain a bit of how it works here. I'm open to a refactor, though.

alexfornuto replied:
I think the volume of explanation here is small enough to warrant a single page. Any more, and we'd want to separate it.

zmb3 commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```
+
+```code
+$ tctl edit users/codingllama
````
First use of `tctl edit` in the docs! 🎉

codingllama replied:
`tctl edit` is too good. I wish we could `tctl edit` resources that don't exist yet...  ;)

codingllama committed:
Review: Explain macOS restriction in text

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+"About This Mac" -> "Serial number").
+
+```code
+$ serial="$(ioreg -c IOPlatformExpertDevice -d 2 | grep -i IOPlatformSerialNumber | awk -F'"' '{print $4}')"
````
Above we tell the reader to replace `$serial` (hence my earlier comment) but here we're explicitly defining it.

Should we make it an uppercase var to follow convention?

codingllama replied:
Yeah, I didn't like this one too much, but it's important to explain what the asset tag is and the command is kinda hard to parse.

I've added a nod to the script in the parenthesis above.

> Should we make it an uppercase var to follow convention?

Done. Do we have our conventions documented somewhere? Browsing other docs for examples is often proving unreliable to me.

alexfornuto replied:
The convention I'm referring to isn't our own.

[Reference 1](https://www.tutorialspoint.com/unix/unix-using-variables.htm#:~:text=The%20name%20of%20a%20variable,have%20their%20names%20in%20UPPERCASE.)

[Reference 2](https://stackoverflow.com/questions/673055/correct-bash-and-shell-script-variable-capitalization)

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+device is enrolled, it becomes a trusted device.
+
+<Admonition type="tip" title="Removing Devices">
+  A device that is no longer in use may be removed using
+  `tctl devices rm --device-id=<ID>` or
+  `tctl devices rm --asset-tag=<SERIAL_NUMBER>`.
```
```suggestion
A device that is no longer in use may be removed using
`tctl devices rm --device-id=<ID>` or
`tctl devices rm --asset-tag=<SERIAL_NUMBER>`.
```

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+
+```code
+$ tctl devices enroll --asset-tag="C00AA3BBBB6C"
+# Run the command below on device "C00AA3BBBB6C" to enroll it:
+# tsh device enroll --token=MQi9gkOfyvdyNyN2moJZz9cqx/GyILR5uNu4NE9dxO8
````
```suggestion
Run the command below on device "C00AA3BBBB6C" to enroll it:
tsh device enroll --token=MQi9gkOfyvdyNyN2moJZz9cqx/GyILR5uNu4NE9dxO8
```

alexfornuto replied:
We should consider if we want to create a preset enrollment token in `config.json` that's obviously fake and use it here and anywhere else we end up talking about device enrollment.

codingllama replied:
Enrollment tokens are random blobs, there's nothing special encoded on them. Happy to use a common token if you think it's best, though.

alexfornuto replied:
Only because we've been told int the past (can't find the issue/PR at the moment) that we get false-positive security reports for having example tokens that look real enough.

codingllama replied:
Done for asset tag and enroll tokens, all files/occurrences.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+# tsh device enroll --token=MQi9gkOfyvdyNyN2moJZz9cqx/GyILR5uNu4NE9dxO8
+```
+
+To perform the enrollment ceremony, using device specified above, type the
````
```suggestion
To perform the enrollment ceremony, using the device specified above, type the
```

codingllama replied:
Done.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```
+
+<Admonition type="warning" title="macOS">
+  macOS enrollments are tied to the current user in the device.
+  If the same device is handed to a new user, they will need to perform
+  enrollment again.
````
```suggestion
macOS enrollments are tied to the current user in the device.
If the same device is handed to a new user, they will need to perform
enrollment again.
```

alexfornuto replied:
IMHO instead of a warning after the reader completes the process, this should be explained at the start of the section, before an admin following along enrolls a device to their own user.

codingllama replied:
Makes sense, moved.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+  <TabItem label="Static Config">
+    Auth Server `teleport.yaml` file:
+
+    ```yaml
+    auth_service:
+      authentication:
+        type: local
+        second_factor: "on"
+        webauthn:
+          rp_id: example.com
+        device_trust:
+          mode: "required" # add this line
+    ```
+  </TabItem>
+  <TabItem label="Dynamic resources">
+    Create a `cap.yaml` file or get the existing configuration using
+    `tctl get cluster_auth_preference`:
+
+    ```yaml
+    kind: cluster_auth_preference
+    version: v2
+    metadata:
+      name: cluster-auth-preference
+    spec:
+      type: local
+      second_factor: "on"
+      webauthn:
+        rp_id: example.com
+      device_trust:
+        mode: "required" # add this line
+    ```
+
+    Update the configuration:
+
+    ```code
+    $ tctl create -f cap.yaml
+    # cluster auth preference has been updated
+    ```
+  </TabItem>
+</Tabs>
````
Edits for copy, indents, and output formatting:

````suggestion
<Tabs>
<TabItem label="Static Config">
Edit the Auth Server's `teleport.yaml` file:

```yaml
auth_service:
  authentication:
    type: local
    second_factor: "on"
    webauthn:
      rp_id: example.com
    device_trust:
      mode: "required" # add this line
```

</TabItem>
<TabItem label="Dynamic resources">
Create a `cap.yaml` file or get the existing configuration using
`tctl get cluster_auth_preference`:

```yaml
kind: cluster_auth_preference
version: v2
metadata:
  name: cluster-auth-preference
spec:
  type: local
  second_factor: "on"
  webauthn:
    rp_id: example.com
  device_trust:
    mode: "required" # add this line
```

Update the configuration:

```code
$ tctl create -f cap.yaml
cluster auth preference has been updated
```

</TabItem>
</Tabs>
````

codingllama replied:
Done.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+For example, SSH access without a trusted device fails with the following error:
+
+```
+$ tsh ssh zarquon
````
Replacing with a more generic example name:

```suggestion
$ tsh ssh server1
```

codingllama replied:
Done.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```
+
+<Admonition type="tip" title="Trusted Clusters">
+  It is possible to use [trusted
+  clusters](../../management/admin/trustedclusters.mdx) to limit the impact of
+  device mode `required`. A leaf cluster in mode `required` will enforce access
+  to all of its resources, without imposing the same restrictions to the root
+  cluster.
````
```suggestion
It is possible to use [trusted
clusters](../../management/admin/trustedclusters.mdx) to limit the impact of
device mode `required`. A leaf cluster in mode `required` will enforce access
to all of its resources, without imposing the same restrictions to the root
cluster.
```

alexfornuto commented on docs/pages/reference/cli.mdx:
````diff
@@ -1096,6 +1096,29 @@ $ tsh request drop [<request-id>...]
+#### Examples
+
+```code
+$ tsh device enroll --token=f3oJYqWfO6XG1ZCIF+faz8HIpDye54LsRFObftGKVVU
+# Device "C00AA3BBBB6C"/macOS enrolled
````
Another spot to replace with an obviously fake token. Maybe we should also create a dummy serial number var as well?

```suggestion
$ tsh device enroll --token=f3oJYqWfO6XG1ZCIF+faz8HIpDye54LsRFObftGKVVU
Device "C00AA3BBBB6C"/macOS enrolled
```

codingllama replied:
That serial number is a dummy, but let me know if you'd rather make it more obvious.

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -2373,6 +2396,113 @@ $ tctl alerts create \
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+| - | - | - | - |
+| `--os` | none | `linux`, `macos`, `windows` | Device operating system |
```
If we're listing all eventual values for `--os`, should we add a note to the reference that only `macos` is currently available?

codingllama replied:
I skipped that on purpose here so there is less risk of forgetting to update it later. The guide does spell it out, as does the command if you try to use it today.

alexfornuto commented on docs/pages/reference/cli.mdx:
````diff
@@ -2373,6 +2396,113 @@ $ tctl alerts create \
     "The system is under maintenance, functionality may be limited."
 ```
 
+### tctl devices add
````
Are these entries being added in any sort of structured order? If not, may I suggest listing them alphabetically?

codingllama replied:
The entries are sorted according to `tctl devices --help`, which uses a logical ordering of sorts: device verbs first (add,ls,rm), then enrollment tokens (enroll). Happy to change as well.

alexfornuto replied:
Following that convention makes sense from that perspective. I'm thinking about the perspective of the user who doesn't have this level of headers listed in our ToC. They may know what they're looking for and be using ctrl/cmd+f to find it, or they may be scrolling the entire document/section to learn. Especially in the latter case, alphabetical ordering can help them keep from losing place.

codingllama replied:
I find this page a bit confusing too, with the lack of navigation and the fact that most of it seems to have no sensible order. Maybe we could add navigation and/or sort the commands consistently throughout? (In a separate PR, that is.)

Sorted `tctl devices` commands alphabetically, but left everything else as-is.

alexfornuto replied:
Thanks! I spun off an issue to tackle the larger doc's sorting.

alexfornuto reviewed (commented):
I reviewed in GitHub only (haven't seen the rendered version), and haven't yet tested.

zmb3 commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
+
+- A signed and notarized `tsh` binary for enrollment.
```
They do need a special version of the binary, because we still ship both signed and unsigned mac binaries (unfortunately).

codingllama replied:
Adding to what Zac said, even the "signed" standalone `tsh` we ship in the tar.gz won't do, as it needs to be in the "tsh.app" shell to work properly. #14006 is the issue meant to address this.

In regards to `tctl`, the include above does cover it, that's why it isn't listed again.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+---
+
+<Admonition type="warning" title="Preview">
+  Device Trust is currently in Preview mode.
```
Done.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+</Admonition>
+
+<ScopedBlock scope="oss">
+  Device Trust requires a commercial edition of Teleport.
```
1. Done. I was mostly following other <Admonition> tags.
2. Done. Ditto, this follows the wording from <Admonitions> in other pages.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+## Concepts
+
+Device Trust uses the Secure Enclave in macOS devices to store a device private
+key. That key is used to solve device challenges, issued by the Teleport Auth
```
done.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+
+Device Trust uses the Secure Enclave in macOS devices to store a device private
+key. That key is used to solve device challenges, issued by the Teleport Auth
+Server, thus proving the identity of the trusted device.
```
Done.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+Device management is divided into two separate phases: inventory management and
+device enrollment.
+
+__Inventory management__ is performed by a device admin. In this step, devices
```
Done throughout.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+
+<Admonition type="tip" title="Enrollment">
+  Device enrollment is key in correctly identifying a device throughout its
+  lifetime. The more trusted the enrollment operator, the better are the ongoing
```
Done. Let me know if you have other suggestions to make it better.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+
+```code
+$ tctl create -f device-admin.yaml
+# role 'device-admin' has been created
````
Done for all code blocks in this file.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
```diff
@@ -0,0 +1,326 @@
+  For production setups a separate role is still recommended.
+</Admonition>
+
+Note that, in addition to the usual CRUD verbs (create, read, list, update and
```
Done.

codingllama committed:
Review: Address Alex's comments

Indents, markdown style, code block style, punctuation and missing
words.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```
+
+```code
+$ tctl edit users/codingllama
````
Done in a way - I show the `tctl edit` then the diff block, but skipped the `tctl edit` output, as it looks odd to repeat the command.

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+
+Make sure your user has the `device-admin` role:
+
+```yaml
````
Good call, done here and for others.

codingllama committed:
Review: Use diff blocks on `tctl edit` example

codingllama committed:
Review: Use diff blocks in other examples

codingllama committed:
Review: Use standard example names

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```yaml
+kind: user
+metadata:
+  name: codingllama
````
Done for all.

codingllama committed:
Review: Sort `tctl devices` commands alphabetically

codingllama committed:
Review: Address Alex's comments (2)

Uppercase SERIAL, macOS enroll notice, config example tabs.

codingllama committed:
Review: Use config variables for asset_tag and enroll_token

codingllama committed:
Review: Remove '#' from code blocks in cli.mdx

codingllama commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,326 @@
+```
+
+Having the necessary permissions, let's add a device to the inventory. Replace
+the `$serial` below with your macOS device serial number (visible under Apple ->
````
Sorry, I'm not sure what you mean here.

codingllama replied:
I gave this a try, but unless I ditch the helper commands below the end result seems worse than now.

codingllama commented:
Friendly ping @alexfornuto ?

alexfornuto commented:
Pings received, will review again today.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,298 @@
+
+```code
+$ SERIAL="$(ioreg -c IOPlatformExpertDevice -d 2 | grep -i IOPlatformSerialNumber | awk -F'"' '{print $4}')"
+$ tctl devices add --os=macos --asset-tag="$SERIAL"
````
It was at this point that my testing failed:

```
❯ tsh version
Teleport v12.0.0-alpha.2 git:api/v12.0.0-alpha.2-0-g3ebfce4a6a go1.19.5
Proxy version: 12.0.0-alpha.2

❯ tctl version
Teleport v12.0.0-alpha.2 git:api/v12.0.0-alpha.2-0-g3ebfce4a6a go1.19.5

❯ tctl get users/alex
kind: user
metadata:
  ...
  name: alex
spec:
  ...
  roles:
  - editor
  - access
  - device-admin
  status:
   ...
version: v2

❯ tctl get roles/device-admin
kind: role
metadata:
  ...
  name: device-admin
spec:
  allow:
    ...
    rules:
    - resources:
      - device
      verbs:
      - create
      - read
      - list
      - update
      - delete
      - create_enroll_token
      - enroll
  deny: {}
  options:
    ...
version: v5

❯ echo $SERIAL
REDACTED

❯ tctl devices add --os=macos --asset-tag="$SERIAL"
tctl: error: expected command but got "devices"

```

codingllama replied:
You'll need the Enterprise version of `tctl` to do this part. We can make that clearer in the docs, or add to the troubleshooting section. WDYT?

Alternatively, we could move this code to OSS `tctl`, say for v12.1. @zmb3 for an opinion.

Commit:
combine warning and ScopedBlock

A ScopedBlock for OSS will never render on a page not scoped for OSS. This way, it's always visible. Once the feature is out of preview, we can just remove the first line.

Commit:
clarify device trust between trusted clusters

This addition dispells any assumptions made by the reader that applying device trust to root clusters would enforece the setting on leaf clusters.

alexfornuto commented on docs/pages/access-controls/guides/device-trust.mdx:
````diff
@@ -0,0 +1,298 @@
+
+```code
+$ SERIAL="$(ioreg -c IOPlatformExpertDevice -d 2 | grep -i IOPlatformSerialNumber | awk -F'"' '{print $4}')"
+$ tctl devices add --os=macos --asset-tag="$SERIAL"
````
I think we can make this more obvious by updating the `/commercial-prereqs-tabs.mdx` codeblock to show the Enterprise output of `tctl version`.

Commit:
update enterprise tctl version output

alexfornuto reviewed (approved):
Tested and confirmed!

```
❯ tsh ssh ubuntu@ip-10-0-1-16
ubuntu@ip-10-0-1-16:~$
```

```
alex@Xeon:~$ tsh ssh ubuntu@ip-10-0-1-16
ERROR: ssh: rejected: administratively prohibited (unauthorized device)
```

I made some minor updates, commit messages should be verbose enough to explain them.

codingllama commented:
Thanks for the reviews and contributions! Merging.
